### PR TITLE
file: add file_system_space

### PR DESCRIPTION
**Pull request by bhalevy**
Return space_info for the filesystem identified
by the given file name.

space_info provides simpler and standard space information about the filesystem, in contrast to the posix statvfs which requires knowledge about the how to convert
f_block to bytes by multiplying by f_frsize.

### CODE REVIEW

File: include/seastar/core/file-types.hh [hunk 1]
```diff
@@ -158,6 +159,18 @@ inline constexpr file_permissions operator&(file_permissions a, file_permissions
     return file_permissions(std::underlying_type_t<file_permissions>(a) & std::underlying_type_t<file_permissions>(b));
 }
 
+/// seastar::space_info is equivalent to std::filesystem::space_info
+/// with renamed members, to prevent a conflict with future::available()
+struct space_info {
```
Comment (avikivity) [inline include/seastar/core/file-types.hh:164]: Why not use std::filesystem::space_info?

File: include/seastar/core/file-types.hh [hunk 2]
```diff
@@ -158,6 +159,18 @@ inline constexpr file_permissions operator&(file_permissions a, file_permissions
...
+    std::uintmax_t capacity;
+    std::uintmax_t free_space;
+    std::uintmax_t available_space;
+
+    bool operator==(const space_info& o) const noexcept {
+        return capacity == o.capacity && free_space == o.free_space && available_space == o.available_space;
+    }
```
Comment (avikivity) [inline include/seastar/core/file-types.hh:171]: = default

File: src/core/reactor.cc [hunk 3]
```diff
@@ -2263,6 +2263,26 @@ reactor::fstatfs(int fd) noexcept {
...
+            sr.throw_fs_exception_if_error("file_system_space failed", pathname);
+            auto& fs_si = sr.extra;
+            return space_info{
+                .capacity = fs_si.capacity,
+                .free_space = fs_si.free,
+                .available_space = fs_si.available
+            };
+        });
+    });
+}
```
Comment (avikivity) [inline src/core/reactor.cc:2284]: Please use coroutines in new code.

File: tests/unit/file_io_test.cc [hunk 4]
```diff
@@ -947,3 +950,20 @@ SEASTAR_TEST_CASE(test_oversized_io_works) {
...
+        space_info si, si1;
+        do {
+            si = file_system_space(name).get();
+            st = engine().statvfs(name).get();
+            si1 = file_system_space(name).get();
+        } while (si1 != si);
+
+        BOOST_REQUIRE_EQUAL(st.f_blocks * st.f_frsize, si.capacity);
+        BOOST_REQUIRE_EQUAL(st.f_bfree * st.f_frsize, si.free_space);
+        BOOST_REQUIRE_EQUAL(st.f_bavail * st.f_frsize, si.available_space);
```
Comment (avikivity) [inline tests/unit/file_io_test.cc:967]: This isn't guaranteed to pass. If a byte is added to the filesystem after assignment to si and removed after assignment to st, si == si1 && si != st.

There isn't a simple way to test this.

Comment (bhalevy) [reply tests/unit/file_io_test.cc:967]: Without syncing the filesystem, it is auto-synced rarely (every 30 seconds?), so we shouldn't see these fluctuations in practice.

File: include/seastar/core/file-types.hh (hunk 1 above)
Comment (bhalevy) [reply include/seastar/core/file-types.hh:164]: I mentioned it in a comment.
The compiler complains about it due to the `available` member conflicting with `future::available`.  I can reproduce and quote the error

File: tests/unit/file_io_test.cc (hunk 4 above)
Comment (avikivity) [reply tests/unit/file_io_test.cc:967]: Who says space is updated on sync?

File: include/seastar/core/file-types.hh (hunk 1 above)
Comment (avikivity) [reply include/seastar/core/file-types.hh:164]: How can it conflict with future::available? It's in a different class.

Comment (avikivity) [reply include/seastar/core/file-types.hh:164]: I see the problem.

Comment (avikivity) [reply include/seastar/core/file-types.hh:164]: https://github.com/scylladb/seastar/pull/2598

File: tests/unit/file_io_test.cc (hunk 4 above)
Comment (bhalevy) [reply tests/unit/file_io_test.cc:967]: Good point, it's probably file system specific.
I saw it with xfs, but tmpfs indeed updates it synchronously.

**Comment by bhalevy** [comment]
Draft until https://github.com/scylladb/seastar/pull/2598

File: src/core/reactor.cc (hunk 3 above)
Comment (bhalevy) [reply src/core/reactor.cc:2275]: Done in v2

**Comment by bhalevy** [comment]
In v2 ([d9b772b](https://github.com/scylladb/seastar/commit/d9b772bb77ef96433764605447ba277106e362dd)):
- rebased on top of #2598 for now
- use std::filesystem::space_info
- use coroutine in reactor::file_system_space
- simplify unit test

File: include/seastar/core/file-types.hh (hunk 2 above)
Comment (bhalevy) [reply include/seastar/core/file-types.hh]: gone in v2

File: include/seastar/core/file-types.hh (hunk 1 above)
Comment (bhalevy) [reply include/seastar/core/file-types.hh]: v2 is rebased onto #2598 for now

File: include/seastar/core/file-types.hh [hunk 5]
```diff
@@ -158,6 +159,18 @@ inline constexpr file_permissions operator&(file_permissions a, file_permissions
     return file_permissions(std::underlying_type_t<file_permissions>(a) & std::underlying_type_t<file_permissions>(b));
 }
 
+/// seastar::space_info is equivalent to std::filesystem::space_info
+/// with renamed members, to prevent a conflict with future::available()
+struct space_info {
+    std::uintmax_t capacity;
+    std::uintmax_t free_space;
+    std::uintmax_t available_space;
```
[the remark this answers is not in the document]
Comment (bhalevy) [reply include/seastar/core/file-types.hh]: gone in v2

**Comment by bhalevy** [comment]
- rebased

**Pull request merged**